### PR TITLE
Edit Article - Link to front-end

### PR DESCRIPTION
**Pull request by pe7er**
Pull Request for Issue #10191 + #10282. This PR adds a **link to the front-end view** of a single article **in the Article: Edit view**. It only works for saved articles (that have an ID) and the links are non-SEF links.
# Testing Instructions
## Before the PR

Go to Content > Articles > open an article. On top it says "Articles: Edit"
administrator-edit_article_before
## After the PR

Go to Content > Articles > open an article. On top it says "Articles: Edit" plus a new icon.
A click on that icon will open the article in the front-end in a new window.
administrator-edit_article_after

### CODE REVIEW

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -396,6 +396,7 @@ JGLOBAL_LINK_AUTHOR_DESC="If set to Yes, the Name of the article's Author will b
 JGLOBAL_LINK_AUTHOR_LABEL="Link Author"
 JGLOBAL_LINK_CATEGORY_DESC="If set to Yes, and if Show Category is set to 'Show', the Category Title will link to a layout showing articles in that Category."
 JGLOBAL_LINK_CATEGORY_LABEL="Link Category"
+JGLOBAL_LINK_FRONTEND_ITEM="Open item in front-end"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Please change to a single word to match the style guide https://joomla.github.io/user-interface-text/?user-interface-text/words2watch.md

Comment (pe7er) [inline administrator/language/en-GB/en-GB.ini]: Thanks @brianteeman ! I've corrected it.
That explains why I could not find "front-end" in the language files :-)

**Comment by pe7er** [comment]
btw: I've used the word "item" in the language file, and not "article".
I'll create two similar PRs for com_contact + com_newsfeeds soon, so we can reuse the language string...

File: administrator/language/en-GB/en-GB.ini (hunk 1 above)
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Thanks

File: administrator/language/en-GB/en-GB.ini [hunk 2]
```diff
@@ -396,6 +396,7 @@ JGLOBAL_LINK_AUTHOR_DESC="If set to Yes, the Name of the article's Author will b
 JGLOBAL_LINK_AUTHOR_LABEL="Link Author"
 JGLOBAL_LINK_CATEGORY_DESC="If set to Yes, and if Show Category is set to 'Show', the Category Title will link to a layout showing articles in that Category."
 JGLOBAL_LINK_CATEGORY_LABEL="Link Category"
+JGLOBAL_LINK_FRONTEND_ITEM="Open item in frontend"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: @infograf768 can confirm but I think the key should be _DESC

Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Yes I noticed that you used item and created a Global key and assumed you were going to be doing more PR ;)

**Comment by brianteeman** [comment]
> It only works for saved articles (that have an ID)

Not 100% true - it only works for saved articles that have a status of published or archived.

Also the icon needs a bit of css love as the colour has little contrast to the background which is an a11y fail and the onhover is almost the same as the background<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by infograf768** [comment]
This PR is wrong as I explained here:
https://github.com/joomla/joomla-cms/pull/10282#issuecomment-228550784

**Comment by infograf768** [comment]
FYI, this is the way we go the correct link using editor-xtd/article
https://github.com/joomla/joomla-cms/commit/6c97f437ec98daeb0e72b8f45e0e3ad369ca5779

**Comment by infograf768** [comment]
Also we get a fatal error
>( ! ) Fatal error: Call to undefined function array_column() in /Applications/MAMP/htdocs/testwindows/trunkgitnew/plugins/system/redirect/redirect.php on line 170

**Comment by infograf768** [comment]
@pe7er 

Here is a diff which will work when sef on or off, in multilang or not.
As JRoute does not work yet from admin to site, it will use a non-sef url.
I guess it is not very important as it is only to check the look of the article in frontend.
The Itemid is the one of the Home page for the language used by the item.
It will be very easy to modify when JRoute will work in admin.
No new lang string necessary as it picks the article title.

```
diff --git a/administrator/components/com_content/views/article/view.html.php b/administrator/components/com_content/views/article/view.html.php
index 9f58183..9e67513 100644
--- a/administrator/components/com_content/views/article/view.html.php
+++ b/administrator/components/com_content/views/article/view.html.php
@@ -87,6 +87,64 @@
 		$canDo = $this->canDo;
 
+		$linkArticle = "";
+
+		if (!$isNew)
+		{
+			 $lang = '';
+			 $sef  = '';
+
+			 // Get the home Itemid for the language
+			$db = JFactory::getDbo();
+			$query = $db->getQuery(true);
+
+			$query->select('id')
+				->from($db->qn('#__menu'))
+				->where($db->qn('home') . '= 1')
+				->where($db->qn('published') . '= 1')
+				->where($db->qn('client_id') . '= 0');
+
+			if (JLanguageMultilang::isEnabled())
+			{
+				$query->where($db->qn('language') . ' = ' . $db->q($this->item->language));
+			}
+			else
+			{
+				$query->where($db->qn('language') . ' = ' . $db->q('*'));
+			}
+			$db->setQuery($query);
+
+			$Itemid = '&amp;Itemid=' . (int) $db->loadResult();
+
+			if ($this->item->language && JLanguageMultilang::isEnabled())
+			{
+				// Get the sef prefix for the language
+				$query->clear()
+					->select('sef')
+					->from($db->qn('#__languages'))
+					->where($db->qn('published') . '= 1')
+					->where($db->qn('lang_code') . ' = ' . $db->q($this->item->language));
+				$db->setQuery($query);
+
+				$sef = $db->loadResult();
+
+				if ($this->item->language != '*')
+				{
+					$lang = '&amp;lang=' . $sef;
+				}
+				else
+				{
+					$lang = "";
+				}
+			}
+
+			$linkArticle = '<a href="' .
+					JURI::root() . 'index.php?option=com_content&amp;view=article&amp;id=' . (int) $this->item->id
+					. '&amp;catid=' . (int) $this->item->catid . $lang . $Itemid . '" title="' . $this->item->title
+					. '" target="_blank"><span class="icon-out-2"></span></a>';
+		}
+
 		JToolbarHelper::title(
-			JText::_('COM_CONTENT_PAGE_' . ($checkedOut ? 'VIEW_ARTICLE' : ($isNew ? 'ADD_ARTICLE' : 'EDIT_ARTICLE'))),
+			JText::_('COM_CONTENT_PAGE_' . ($checkedOut ? 'VIEW_ARTICLE' : ($isNew ? 'ADD_ARTICLE' : 'EDIT_ARTICLE')))
+			. " " . $linkArticle,
 			'pencil-2 article-add'
 		);
```

**Comment by mbabker** [comment]
@infograf768 log a separate (release blocking) issue for the `array_column()` error.  It's unrelated to this patch and it is a release blocker because [per the documentation](http://php.net/manual/en/function.array-column.php) it only exists as of PHP 5.5.  Either a polyfill for the method needs to be implemented (add the `symfony/polyfill-php55` package via Composer) or the code rewritten to not use that function.

**Comment by infograf768** [comment]
@mbabker 
Done

**Comment by C-Lodder** [comment]
Good idea but I am not a fan of the approach. Having a single icon next to the Toolbar title has no real meaning so people will not know what it does.

I think adding it below the textarea as show in the screenshot below would be better. The icon for the "toggle editor" button also really needs to change too:

screeny

**Comment by ggppdk** [comment]
That looks like you are previewing only description ?

but also i agree that an icon next to "Page title" is best choice

- there should be a new button next to the buttions "save" , "close" etc

**Comment by brianteeman** [comment]
I agree about the icon - I will see what I can find (for another PR)

-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by killoltailored** [comment]
I have tested this item :white_check_mark: successfully on b8805e7b3116fc1205cbd56b72cefd7f92fec2da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by TasolJamal** [comment]
I have tested this item :white_check_mark: successfully on b8805e7b3116fc1205cbd56b72cefd7f92fec2da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

File: administrator/components/com_content/views/article/view.html.php [hunk 3]
```diff
@@ -86,8 +86,19 @@ protected function addToolbar()
 		// Built the actions for new and existing records.
 		$canDo = $this->canDo;
 
+		$linkArticle = "";
+
+		if (!$isNew)
+		{
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/article/view.html.php]: wouldn't this be more correct?
```php
if (!$isNew)
{
	JLoader::register('ContentHelperRoute', JPATH_SITE . '/components/com_content/helpers/route.php');
	$url = JRoute::_(JURI::root() . ContentHelperRoute::getArticleRoute((int) $this->item->id, (int) $this->item->catid, $this->item->language));
	$linkArticle = '<a href="' . $url . '" title="' . JText::_('JGLOBAL_LINK_FRONTEND_ITEM') . '" target="_blank"><span class="icon-out-2"></span></a>';
}
```

And if/when https://github.com/joomla/joomla-cms/pull/11060 get merged

```php
if (!$isNew)
{
	JLoader::register('ContentHelperRoute', JPATH_SITE . '/components/com_content/helpers/route.php');
	$url = JRoute::_(ContentHelperRoute::getArticleRoute((int) $this->item->id, (int) $this->item->catid, $this->item->language), true, null, 'site');
	$linkArticle = '<a href="' . $url . '" title="' . JText::_('JGLOBAL_LINK_FRONTEND_ITEM') . '" target="_blank"><span class="icon-out-2"></span></a>';
}
```

**Comment by gunjanpatel** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by brianteeman** [comment]
Sorry I am removing the RTC at this time. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by infograf768** [comment]
Indeed. Test were not done correctly. Once more, the language has to be included in the link... !!! See above.

**Comment by infograf768** [comment]
Once https://github.com/joomla/joomla-cms/issues/11060 is merged, the right code would be:
```
diff --git a/administrator/components/com_content/views/article/view.html.php b/administrator/components/com_content/views/article/view.html.php
index 9f58183..68fdc18 100644
--- a/administrator/components/com_content/views/article/view.html.php
+++ b/administrator/components/com_content/views/article/view.html.php
@@ -87,6 +87,64 @@
 		$canDo = $this->canDo;
 
+		$linkArticle = "";
+
+		if (!$isNew)
+		{
+			 $lang = '';
+			 $sef  = '';
+
+			 // Get the home Itemid for the language
+			$db = JFactory::getDbo();
+			$query = $db->getQuery(true);
+
+			$query->select('id')
+				->from($db->qn('#__menu'))
+				->where($db->qn('home') . '= 1')
+				->where($db->qn('published') . '= 1')
+				->where($db->qn('client_id') . '= 0');
+
+			if (JLanguageMultilang::isEnabled())
+			{
+				$query->where($db->qn('language') . ' = ' . $db->q($this->item->language));
+			}
+			else
+			{
+				$query->where($db->qn('language') . ' = ' . $db->q('*'));
+			}
+			$db->setQuery($query);
+
+			$Itemid = '&amp;Itemid=' . (int) $db->loadResult();
+
+			if ($this->item->language && JLanguageMultilang::isEnabled())
+			{
+				// Get the sef prefix for the language
+				$query->clear()
+					->select('sef')
+					->from($db->qn('#__languages'))
+					->where($db->qn('published') . '= 1')
+					->where($db->qn('lang_code') . ' = ' . $db->q($this->item->language));
+				$db->setQuery($query);
+
+				$sef = $db->loadResult();
+
+				if ($this->item->language != '*')
+				{
+					$lang = '&amp;lang=' . $sef;
+				}
+				else
+				{
+					$lang = "";
+				}
+			}
+
+			$link = JRoute::_('index.php?option=com_content&amp;view=article&amp;id=' . (int) $this->item->id
+					. '&amp;catid=' . (int) $this->item->catid . $lang . $Itemid , true, null, 'site');
+			$linkArticle = '<a href="' . $link . '" title="' . $this->item->title
+					. '" target="_blank"><span class="icon-out-2"></span></a>';
+		}
+
 		JToolbarHelper::title(
-			JText::_('COM_CONTENT_PAGE_' . ($checkedOut ? 'VIEW_ARTICLE' : ($isNew ? 'ADD_ARTICLE' : 'EDIT_ARTICLE'))),
+			JText::_('COM_CONTENT_PAGE_' . ($checkedOut ? 'VIEW_ARTICLE' : ($isNew ? 'ADD_ARTICLE' : 'EDIT_ARTICLE')))
+			. " " . $linkArticle,
 			'pencil-2 article-add'
 		);
```

**Comment by infograf768** [comment]
@andrepereiradasilva 
As you see in my code above, I did not need
` JRoute::_(ContentHelperRoute::getArticleRoute`

**Comment by andrepereiradasilva** [comment]
> @andrepereiradasilva 
> As you see in my code above, I did not need
> JRoute::_(ContentHelperRoute::getArticleRoute

true, but if already exists why not just reuse it like i did in my comment?

**Comment by infograf768** [comment]
Indeed. :)

**Comment by infograf768** [comment]
This would be enough:|
```
		$linkArticle = "";
		$url         = "";

		if (!$isNew)
		{
			JLoader::register('ContentHelperRoute', JPATH_SITE . '/components/com_content/helpers/route.php');
    		$url = JRoute::_(ContentHelperRoute::getArticleRoute((int) $this->item->id, (int) $this->item->catid, $this->item->language), true, null, 'site');
    		$linkArticle = '<a href="' . $url . '" title="' . htmlspecialchars($this->item->title, ENT_COMPAT, 'UTF-8') . '" target="_blank"><span class="icon-out-2"></span></a>';
		}

		JToolbarHelper::title(
			JText::_('COM_CONTENT_PAGE_' . ($checkedOut ? 'VIEW_ARTICLE' : ($isNew ? 'ADD_ARTICLE' : 'EDIT_ARTICLE')))
			. " " . $linkArticle,
			'pencil-2 article-add'
		);
```

**Comment by NeerajTailored** [comment]
I have tested this item :white_check_mark: successfully on b8805e7b3116fc1205cbd56b72cefd7f92fec2da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by brianteeman** [comment]
@pe7er will you be updating this to work with multilingual as mentioned by @infograf768 above?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/joomla-cms/10932</a>.</sub>

**Comment by pe7er** [comment]
Actually, I prefer @C-Lodder his approach, with adding a button below the article.
Could you please create a PR for your suggestion?

**Comment by C-Lodder** [comment]
I can do tonight. Would it not be best to submit a PR against the 3.7 branch so that https://github.com/joomla/joomla-cms/pull/11060 will be merged too?

**Comment by brianteeman** [comment]
@c-lodder any update?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/tracker/joomla-cms/10932</a>.</sub>

**Comment by C-Lodder** [comment]
@brianteeman - will do now. So I assume this is going into staging then?

**Comment by C-Lodder** [comment]
PR: https://github.com/joomla/joomla-cms/pull/12367

**Comment by brianteeman** [comment]
@pe7er not sure what the status is on this PR - can you clarify please<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/tracker/joomla-cms/10932</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on b8805e7b3116fc1205cbd56b72cefd7f92fec2da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10932">issues.joomla.org/tracker/joomla-cms/10932</a>.</sub>

**Comment by AndySDH** [comment]
How comes this was never implemented? Seems like a no-brainer.

**Comment by C-Lodder** [comment]
@AndySDH - It's already in J4.x (display in a modal) so people are more than welcome to copy it into 3.x

**Comment by franz-wohlkoenig** [comment]
@C-Lodder haven't seen the new "Preview"-Button before.

**Comment by C-Lodder** [comment]
@franz-wohlkoenig https://gyazo.com/3f79759488413cc5acd3d6f0daf479ad

**Comment by franz-wohlkoenig** [comment]
@C-Lodder i meant: Havent seen before your [comment](https://github.com/joomla/joomla-cms/pull/10932#issuecomment-281613724);-)

**Comment by roland-d** [comment]
@pe7er I am going to close this issue as it is already implemented in Joomla 4 and Joomla 4 is on the horizon. This may not be worth the effort.

**Comment by pe7er** [comment]
Ok, thanks @roland-d

**Comment by 4digitalworld** [comment]
@roland-d @pe7er Hello from 2020! Almost 3 years have passed, and we still do not see this function, just as we do not even see the beta version of Joomla 4.

How many articles could be created using this feature over the last 3 years...

**Pull request closed without merging**